--- Day 6: Trash Compactor ---
After helping the Elves in the kitchen, you were taking a break and helping them re-enact a movie scene when you over-enthusiastically jumped into the garbage chute!

A brief fall later, you find yourself in a garbage smasher. Unfortunately, the door's been magnetically sealed.

As you try to find a way out, you are approached by a family of cephalopods! They're pretty sure they can get the door open, but it will take some time. While you wait, they're curious if you can help the youngest cephalopod with her math homework.

In [30]:
elf_data = []

with open('day06.txt', 'r') as file:
    for line in file:
        elf_data.append(line.strip())

print(len(elf_data))
print(elf_data[0:10])
print(type(elf_data))


5
['8 797  34  5 84 14  57 9   831 5934 639 888  9 12 6421 825 8418 9   6  21  54 37  95 138   17 778 63 88   9 41  1 731 653 544 9394 48 32 939 995 52 1966 98  834  71   7 42 995  9 9  223 7  929 4  99 2161  68 729 171 665 227 81  71  68 61 4892 34 26 11  152 5   2763 41 157 3   83 91 132 3352 33 1  69  16  331  4 89 43 258 12 561 29 63 23 583 53  176  14  7373 887  491 582  4 123 527  9  28 47  4196 74  82 328 1478 13 7341 694 921 81   5 81 76  1 291 1543 35 446 829 579 51 73 232 65 48 5    47  9 142 54  8 85 3   84 81 77 716    1  15 36 4  112 285 398 2   6955 843 141  39 48 3485   7 267 6898 1699 24 5163 95   36 7   9   378  88 88 6    6  732 48 1    72 514  647 818 14 2741 11 6425 568 77 16 133  5 799 599 228  11  26 3  193 44 83 73 5619 68 3   514 1895 793 9  5135 621 914  81 389 87 42   579 7249  9 946 32 2    331 9  5  64 94  549 52 9   4554 45   766 28 396  461  3  5 71 1   288 561  282 16 757 5535 771  175  734 2431   4 216 54  16  33  73 32 23   3 6  469 99 1111 3878 76 51  

Cephalopod math doesn't look that different from normal math. The math worksheet (your puzzle input) consists of a list of problems; each problem has a group of numbers that need to either be either added (+) or multiplied (*) together.

However, the problems are arranged a little strangely; they seem to be presented next to each other in a very long horizontal list. For example:

123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  

Each problem's numbers are arranged vertically; at the bottom of the problem is the symbol for the operation that needs to be performed. Problems are separated by a full column of only spaces. The left/right alignment of numbers within each problem can be ignored.

So, this worksheet contains four problems:

123 * 45 * 6 = 33210
328 + 64 + 98 = 490
51 * 387 * 215 = 4243455
64 + 23 + 314 = 401

In [31]:
test_data = ["123 328  51 64", 
"45 64  387 23", 
"6 98  215 314",
"*   +   *   +"]

test_data

['123 328  51 64', '45 64  387 23', '6 98  215 314', '*   +   *   +']

In [32]:
import numpy as np
import regex as re

In [33]:
for i in range(0, len(test_data)):
    test_data[i] = re.split(r' +', test_data[i])

test_data

[['123', '328', '51', '64'],
 ['45', '64', '387', '23'],
 ['6', '98', '215', '314'],
 ['*', '+', '*', '+']]

In [34]:
for i in range(0, len(elf_data)):
    elf_data[i] = re.split(r' +', elf_data[i])

#elf_data[:-1]

[['8',
  '797',
  '34',
  '5',
  '84',
  '14',
  '57',
  '9',
  '831',
  '5934',
  '639',
  '888',
  '9',
  '12',
  '6421',
  '825',
  '8418',
  '9',
  '6',
  '21',
  '54',
  '37',
  '95',
  '138',
  '17',
  '778',
  '63',
  '88',
  '9',
  '41',
  '1',
  '731',
  '653',
  '544',
  '9394',
  '48',
  '32',
  '939',
  '995',
  '52',
  '1966',
  '98',
  '834',
  '71',
  '7',
  '42',
  '995',
  '9',
  '9',
  '223',
  '7',
  '929',
  '4',
  '99',
  '2161',
  '68',
  '729',
  '171',
  '665',
  '227',
  '81',
  '71',
  '68',
  '61',
  '4892',
  '34',
  '26',
  '11',
  '152',
  '5',
  '2763',
  '41',
  '157',
  '3',
  '83',
  '91',
  '132',
  '3352',
  '33',
  '1',
  '69',
  '16',
  '331',
  '4',
  '89',
  '43',
  '258',
  '12',
  '561',
  '29',
  '63',
  '23',
  '583',
  '53',
  '176',
  '14',
  '7373',
  '887',
  '491',
  '582',
  '4',
  '123',
  '527',
  '9',
  '28',
  '47',
  '4196',
  '74',
  '82',
  '328',
  '1478',
  '13',
  '7341',
  '694',
  '921',
  '81',
  '5',
  '81',
  '76',
  '1',

In [37]:
np_test_data = np.array(test_data)
np_elf_data = np.array(elf_data)
np_test_data

array([['123', '328', '51', '64'],
       ['45', '64', '387', '23'],
       ['6', '98', '215', '314'],
       ['*', '+', '*', '+']], dtype='<U3')

In [38]:
tdt = np.transpose(np_test_data)
edt = np.transpose(np_elf_data)
tdt

array([['123', '45', '6', '*'],
       ['328', '64', '98', '+'],
       ['51', '387', '215', '*'],
       ['64', '23', '314', '+']], dtype='<U3')

To check their work, cephalopod students are given the grand total of adding together all of the answers to the individual problems. In this worksheet, the grand total is 33210 + 490 + 4243455 + 401 = 4277556.

In [51]:
def calculate(data):
    total = 0
    for d in data:
        print(d)
        if d[-1] == "+":
            total += adding(d[:-1])
        else:
            if d[-1] == "*":
                total += multi(d[:-1])
        print(d, total)
    return total

def adding(numbers):
    total = 0
    for n in numbers:
        total = total+int(n)
    print(total, "adding ", numbers)        
    return total

def multi(numbers):
    total = 1
    for n in numbers:
        total = total * int(n)
    print(total, "multip ", numbers)
    return total

calculate(tdt)

['123' '45' '6' '*']
33210 multip  ['123' '45' '6']
['123' '45' '6' '*'] 33210
['328' '64' '98' '+']
490 adding  ['328' '64' '98']
['328' '64' '98' '+'] 33700
['51' '387' '215' '*']
4243455 multip  ['51' '387' '215']
['51' '387' '215' '*'] 4277155
['64' '23' '314' '+']
401 adding  ['64' '23' '314']
['64' '23' '314' '+'] 4277556


4277556

In [42]:
tdt[:-1]

array([['123', '45', '6', '*'],
       ['328', '64', '98', '+'],
       ['51', '387', '215', '*']], dtype='<U3')

Of course, the actual worksheet is much wider. You'll need to make sure to unroll it completely so that you can read the problems clearly.

Solve the problems on the math worksheet. What is the grand total found by adding together all of the answers to the individual problems?


In [52]:
calculate(edt)

['8' '75' '23' '29' '*']
400200 multip  ['8' '75' '23' '29']
['8' '75' '23' '29' '*'] 400200
['797' '142' '395' '21' '+']
1355 adding  ['797' '142' '395' '21']
['797' '142' '395' '21' '+'] 401555
['34' '794' '753' '161' '+']
1742 adding  ['34' '794' '753' '161']
['34' '794' '753' '161' '+'] 403297
['5' '58' '59' '94' '*']
1608340 multip  ['5' '58' '59' '94']
['5' '58' '59' '94' '*'] 2011637
['84' '45' '42' '86' '+']
257 adding  ['84' '45' '42' '86']
['84' '45' '42' '86' '+'] 2011894
['14' '55' '22' '65' '+']
156 adding  ['14' '55' '22' '65']
['14' '55' '22' '65' '+'] 2012050
['57' '811' '979' '233' '+']
2080 adding  ['57' '811' '979' '233']
['57' '811' '979' '233' '+'] 2014130
['9' '965' '123' '687' '*']
733891185 multip  ['9' '965' '123' '687']
['9' '965' '123' '687' '*'] 735905315
['831' '634' '37' '6' '*']
116961588 multip  ['831' '634' '37' '6']
['831' '634' '37' '6' '*'] 852866903
['5934' '9431' '5319' '34' '+']
20718 adding  ['5934' '9431' '5319' '34']
['5934' '9431' '5319' '34' 

5381996914800


--- Part Two ---
The big cephalopods come back to check on how things are going. When they see that your grand total doesn't match the one expected by the worksheet, they realize they forgot to explain how to read cephalopod math.

Cephalopod math is written right-to-left in columns. Each number is given in its own column, with the most significant digit at the top and the least significant digit at the bottom. (Problems are still separated with a column consisting only of spaces, and the symbol at the bottom of the problem is still the operator to use.)

Here's the example worksheet again:

123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
Reading the problems right-to-left one column at a time, the problems are now quite different:

The rightmost problem is 4 + 431 + 623 = 1058
The second problem from the right is 175 * 581 * 32 = 3253600
The third problem from the right is 8 + 248 + 369 = 625
Finally, the leftmost problem is 356 * 24 * 1 = 8544
Now, the grand total is 1058 + 3253600 + 625 + 8544 = 3263827.

Solve the problems on the math worksheet again. What is the grand total found by adding together all of the answers to the individual problems?